### ISSUE TITLE
Rule: VariableDeclarationAfterUsage gives false positives

### ISSUE THREAD

**Issue by BasieP** [SonarSource/SonarJS#808]
the rule VariableDeclarationAfterUsage gives lots of false positives in our code.

we have stuff like this:

```javascript
const func1 = function() {
    return 'foo' + func2();
};
const func2 = function() {   
    if (someCondition) {
        return 'bar';
    } else {
        return func1();
    }
};

func1();
```

this is perfectly fine code. First we declare both functions. This makes that the fact that the functions call each other is wrong in no possible way.

i don't redeclare the variable (func1 or func2) in other scopes (like this example page shows https://github.com/SonarSource/SonarJS/blob/master/javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/VariableDeclarationAfterUsage.html)

So can this rule please be make a little bit smarter? function declaration is chronologically way before the execution of the function.

**Comment by vilchik-elena** [comment]
@BasieP Why you don't declare function as function declaration? Your code is really not safe. If later on you refactor your code to 
```
 const func1 = function() {
    return 'foo' + func2();
};
func1();
const func2 = function() {   
    if (someCondition) {
        return 'bar';
    } else {
        return func1();
    }
};
```
it will fail with type error.

With function declaration issue disappears.

**Comment by BasieP** [comment]
I use Javascript that should work on ie11, so i don't know how your
'function declaration' is, but this is valid js code.

If i refactor my code it should fail sonar test THEN. Not now.
So i call function func1 which calls func2 which at that point is not
defined yet.
Yes it should fail.

But in my first example it should not.



On 26 Oct 2017 4:55 pm, "Elena Vilchik" <notifications@github.com> wrote:

@BasieP <https://github.com/basiep> Why you don't declare function as
function declaration? Your code is really not safe. If later on you
refactor your code to

 const func1 = function() {
    return 'foo' + func2();
};
func1();
const func2 = function() {
    if (someCondition) {
        return 'bar';
    } else {
        return func1();
    }
};

it will fail with type error.

With function declaration issue disappears.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/SonarSource/SonarJS/issues/808#issuecomment-339693031>,
or mute the thread

**Comment by vilchik-elena** [comment]
@BasieP I see your point, indeed it would be perfect to not raise issue in first case, and do it only when error will occur for sure. The point is that it's really hard to implement it that way. But of course, if there will be more feedback for this rule, we will reconsider it.

In the meantime, you can disable the rule or use function declarations like this:
```javascript
function func1 () {
    return 'foo' + func2();
}
function func2() {
    if (someCondition) {
        return 'bar';
    } else {
        return func1();
    }
}
```
Function declaration should perfectly work from ie11, the difference from your code, that function implementations will be [hoisted](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting) before any other code is executed.

**Comment by BasieP** [comment]
I don't know a lot about the sonar rules and how they are build up, but i'm sure it can be done by simply building a stack of calls.
How can i learn how these rules are applied?

**Comment by vilchik-elena** [comment]
@BasieP You are welcome to provide PR if you see how simple it is:) 
But honestly saying, it might be simple in your trivial case, but it's not in general case due to dynamic nature of JS. 
Anyway you can find some docs for writing rules [here](https://docs.sonarqube.org/display/PLUG/Custom+Rules+for+SonarJS) and of course all our rules in this repo are at your disposal.

**Comment by ptcc** [comment]
eslint gets this right
